pkg/vsphere/compute/placement/placement.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2018 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package placement
+
+import (
+	"fmt"
+	"math/rand"
+	"sort"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/compute"
+	"github.com/vmware/vic/pkg/vsphere/performance"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+)
+
+const (
+	memUnconsumedWeight = 0.7 // available memory (total - consumed)
+	memInactiveWeight   = 0.3 // active memory on the host
+)
+
+type rankedHost struct {
+	*object.HostSystem
+	*performance.HostMetricsInfo
+	score float64
+}
+
+func (r *rankedHost) String() string {
+	return fmt.Sprintf("%s:%f", r.HostSystem.InventoryPath, r.score)
+}
+
+type rankedHosts []rankedHost
+
+func (r rankedHosts) Len() int           { return len(r) }
+func (r rankedHosts) Swap(i, j int)      { r[i], r[j] = r[j], r[i] }
+func (r rankedHosts) Less(i, j int) bool { return r[i].score > r[j].score }
+
+// Engine uses data from a MetricsProvider to decide on which host to powerOn a VM.
+type Engine struct {
+	source performance.MetricsProvider
+}
+
+// NewEngine returns an Engine instance using the supplied MetricsProvider.
+func NewEngine(s performance.MetricsProvider) *Engine {
+	return &Engine{source: s}
+}
+
+// CheckHost returns true if the host has adequate capacity to power on the VM, false otherwise.
+func (e *Engine) CheckHost(op trace.Operation, host *object.HostSystem) bool {
```

Comment:
@hickeng I like the idea of using input types from the same package, but if we switch to `*object.VirtualMachine`, we lose access to the `session.Session` which is required to find the compute resource. Any way around that?

Reply:
I've gone ahead and defined the `HostPlacementPolicy` interface, and implemented it in the form of `RandomHostPolicy`. I've left the input types as they are until a better design is discovered around how we obtain the compute resource via the `Session`.

After a discussion with @hickeng, we've decided to merge this as is, and further work addressing the weighting configuration and ranking algorithm, as well as tests using the VC simulator, will come in a followup PR.

Reply:
I don't really like the placement logic needing to hold a `Session` just in order to get the cluster either. Correct answer is probably to move that call into govmomi so it can use the object.VirtualMachine.client. Keep it as `vm.VirtualMachine` for now.